packages/flutter/lib/src/cupertino/text_field.dart

```diff
@@ -0,0 +1,600 @@
+    final FocusNode focusNode = _effectiveFocusNode;
+    final List<TextInputFormatter> formatters = widget.inputFormatters ?? <TextInputFormatter>[];
+    final bool enabled = widget.enabled ?? true;
+    if (widget.maxLength != null && widget.maxLengthEnforced)
+      formatters.add(LengthLimitingTextInputFormatter(widget.maxLength));
+
+    Widget child = Padding(
+      padding: widget.padding,
+      child: RepaintBoundary(
+        child: EditableText(
+          key: _editableTextKey,
+          controller: controller,
+          focusNode: focusNode,
+          keyboardType: widget.keyboardType,
+          textInputAction: widget.textInputAction,
+          textCapitalization: widget.textCapitalization,
+          style: widget.style,
+          textAlign: widget.textAlign,
+          autofocus: widget.autofocus,
+          obscureText: widget.obscureText,
+          autocorrect: widget.autocorrect,
+          maxLines: widget.maxLines,
+          selectionColor: _kSelectionHighlightColor,
+          selectionControls: cupertinoTextSelectionControls,
+          onChanged: widget.onChanged,
+          onEditingComplete: widget.onEditingComplete,
+          onSubmitted: widget.onSubmitted,
+          inputFormatters: formatters,
+          rendererIgnoresPointer: true,
+          cursorWidth: widget.cursorWidth,
+          cursorRadius: widget.cursorRadius,
+          cursorColor: widget.cursorColor,
+          scrollPadding: widget.scrollPadding,
+          keyboardAppearance: keyboardAppearance,
+        ),
+      ),
+    );
+
+    if (widget.placeholder != null ||
+        widget.clearButtonMode != OverlayVisibilityMode.never ||
+        widget.prefix != null ||
+        widget.suffix != null) {
+      child = ValueListenableBuilder<TextEditingValue>(
+        valueListenable: controller,
+        builder: (BuildContext context, TextEditingValue text, Widget child) {
+          final List<Widget> rowChildren = <Widget>[];
+
+          if (widget.prefix != null &&
+              _shouldAppearForModeAndText(widget.prefixMode, text.text.isNotEmpty)) {
+            rowChildren.add(widget.prefix);
+          }
+
+          final List<Widget> stackChildren = <Widget>[];
+
+          if (widget.placeholder != null && text.text.isEmpty) {
+            stackChildren.add(
+              Padding(
+                padding: widget.padding,
+                child: Text(
+                  widget.placeholder,
+                  maxLines: 1,
+                  style: widget.style.merge(
+                    const TextStyle(
+                      color: _kInactiveTextColor,
+                      fontWeight: FontWeight.w300,
+                    ),
+                  ),
+                ),
+              ),
+            );
+          }
+
+          rowChildren.add(Expanded(child: Stack(children: stackChildren..add(child))));
+
+          if (widget.suffix != null &&
+              _shouldAppearForModeAndText(widget.suffixMode, text.text.isNotEmpty)) {
```

Comment:
Ah, I see, this method takes in many different things.  I would definitely break this method apart.  As a reader, I'd like to fully understand the nature of what it is being called, but methods that essentially take in switchable values require a developer to jump into the method to understand what it's really used for.

So if you want to ask the question "should we show the prefix", then I'd like to see a method like _shouldShowPrefix(...), and then _shoudlShowSuffix(...).  That greatly reduces the cognitive overhead of reading through the code.

Reply:
Ya, I don't have a super great solution. Though more indirections impedes readability too. Hopefully the compromise is reasonable by shortening the method and highlighting the key part more quickly with named params.

Reply:
Never mind. Did what you suggested